Add DAC emulator driver

lindblandro opened the pull request:
This driver allows for reading back the latest written value for a channel in tests. The design is mimicked from the "adc-emul" driver.

vai-helin committed:
drivers: dac: add emulator driver

For use in tests where the existing "vnd,dac" driver doesn't work as it
returns error codes for all API functions. The approach is mimicked from
the adc-emul driver.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

vai-helin committed:
tests: drivers: dac: add dac-emul to build_all

Add the new dac emulator driver to the build-all testsuite.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

martinjaeger commented on tests/drivers/dac/dac_emul/src/main.c:
```diff
@@ -0,0 +1,589 @@
+static const struct device *dac_emul_many = DEVICE_DT_GET(DAC_EMUL_MANY_NODE);
+
+/* External API for testing */
+extern int dac_emul_value_get(const struct device *dev, uint8_t channel, uint32_t *value);
```
Why not include `zephyr/drivers/dac/dac_emul.h` instead?

martinjaeger commented on tests/drivers/dac/dac_emul/src/main.c:
```diff
@@ -0,0 +1,589 @@
+	ret = dac_emul_value_get(dac_emul0, 0, &value);
+	zassert_equal(ret, 0, "Read should succeed");
+	zassert_equal(value, 512, "Read value should match");
+}
```
As far as I can see, this does exactly the same as `test_write_value_basic`. I suggest to keep only one of the two.

martinjaeger reviewed (changes requested):
The test cases are very extensive and I'm not sure if that's really required. Can you please explain the rationale behind all these tests? Are we not just checking each and every corner of a driver which is only for testing and will not even be used in any product at any time?

lindblandro commented:
> The test cases are very extensive and I'm not sure if that's really required. Can you please explain the rationale behind all these tests? Are we not just checking each and every corner of a driver which is only for testing and will not even be used in any product at any time?

To be honest, I mainly added these as I saw that there were tests for `adc_emul`.  Those probably make more sense as it's the beginning of the signal processing chain and the ADC interface has more features. In any case I'll prune and simplify these tests.

martinjaeger reviewed (dismissed):
I was wondering if we should move some of the tests to the existing `dac_api` test, but I realized that most tests for the emul read the values back (which would not work with actual HW). So let's keep like this for now.

Thanks for your patience.

petejohanson-adi reviewed (changes requested):
A few thoughts. Thanks!

vai-helin committed:
tests: drivers: dac_emul: add ztests

Add unit tests that stress the driver in various ways. The concurrent
access test is a best effort attempt to trigger race conditions.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

lindblandro commented on drivers/dac/dac_emul.c:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2025 Vaisala Oyj
```
They were accurate when the PR was opened, but sure :)

petejohanson-adi commented on drivers/dac/dac_emul.c:
Probably should be returning the `rc` value from two lines up?

petejohanson-adi reviewed (changes requested):
Just one minor thing after the refactor for the checks, otherwise LGTM.
